### PR TITLE
#1984 Sell and cost columns

### LINKED ISSUE #1984: Stocktake and SI sell and cost prices when payments enabled
## Is your feature request related to a problem? Please describe.

When payments are enabled, the sell and cost prices of items must be available to be entered for incoming stock

## Describe the solution you'd like

N/A

## Implementation

Add `sellPrice` and `costPrice` columns to SI and stocktake edit batch modal

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #1984 

## Change summary

- Adds sell and cost price columns to SI's and stocktake batch modal:

image

image

- Cost is not editable sell price is - both will edit the lines underlying item batches sell price as well as the stocktake batch or transaction batch
- Editing isn't ideal as you for example if you edit $10.00 -> $1000 this becomes $1,000.00 ... I think best to use a text input mask rather than typing to a `currency` - but this would probably mean making a `CurrencyInputCell` which I can do a little later...!

## Testing

N/A

### Related areas to think about

- Still not really sure how dispensing per item rather than batches with multiple different prices is really going to work.. I suppose can list a range of prices the batches have.. and then list the batches separately at the end? I guess!

### CODE REVIEW

File: src/widgets/modals/StocktakeBatchModal.js [hunk 1]
```diff
@@ -185,8 +186,24 @@ export const StocktakeBatchModal = ({ stocktakeItem }) => {
...
+};
+
+const mapStateToProps = state => {
+  const usingPaymentsModule = selectUsingPayments(state);
+  const usingReasons =
+    UIDatabase.objects('NegativeAdjustmentReason').length > 0 &&
+    UIDatabase.objects('PositiveAdjustmentReason').length > 0;
+
+  if (usingReasons) {
+    if (usingPaymentsModule) return { page: 'stocktakeBatchEditModalWithReasonsAndPrices' };
```
Comment (wlthomson) [inline src/widgets/modals/StocktakeBatchModal.js]: Just a suggestion as know this is a WIP, but could maybe clean up the long strings with page getters? E.g.

```
page = getPage('stocktakeBatch', 'editModal', { usingReasons, usingPrices });
```

Definitely OTT and would be hard to design elegantly, but a sort of "higher-order" factory/constructor is also a cool option :smiley: , e.g.

```
page = getPage( 'stocktakeBatch', 'editModal').with({ usingReasons, usingPrices });
```

File: src/database/DataTypes/StocktakeBatch.js [hunk 2]
```diff
@@ -267,6 +268,14 @@ export class StocktakeBatch extends Realm.Object {
   toString() {
     return `Stocktake batch representing ${this.itemBatch}`;
   }
+
+  get costPriceString() {
+    return currency(this.costPrice || 0, { formatWithSymbol: true }).format();
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/StocktakeBatch.js]: 5c244307e1782dad13d2c3fb72ae6dfedaf7e9c3

File: src/pages/dataTableUtilities/actions/cellActions.js [hunk 3]
```diff
@@ -53,6 +54,25 @@ export const editBatchName = (value, rowKey, objectType, route) => (dispatch, ge
...
 
+export const editSellPrice = (value, rowKey, route) => (dispatch, getState) => {
+  const { data, keyExtractor } = pageStateSelector(getState());
+
+  const objectToEdit = data.find(row => keyExtractor(row) === rowKey);
+  const objectDataType = objectToEdit.stocktake ? 'StocktakeBatch' : 'TransactionBatch';
+
+  const { itemBatch, sellPrice } = objectToEdit;
+
+  const valueAsCurrency = currency(value);
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/dataTableUtilities/actions/cellActions.js]: 83ac79ead52cd76d6506e7ee543c25b60bbe4f5c

File: src/widgets/modals/StocktakeBatchModal.js [hunk 4]
```diff
@@ -185,8 +186,24 @@ export const StocktakeBatchModal = ({ stocktakeItem }) => {
...
+StocktakeBatchModalComponent.propTypes = {
   stocktakeItem: PropTypes.object.isRequired,
+  page: PropTypes.string.isRequired,
+};
+
+const mapStateToProps = state => {
+  const usingPaymentsModule = selectUsingPayments(state);
+  const usingReasons =
+    UIDatabase.objects('NegativeAdjustmentReason').length > 0 &&
+    UIDatabase.objects('PositiveAdjustmentReason').length > 0;
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/widgets/modals/StocktakeBatchModal.js]: the + and - got added only recently. The logic would get even more annoying and complicated than it already is if we didn't enforce both existing

File: src/widgets/modals/StocktakeBatchModal.js (hunk 1 above)
Comment (joshxg) [reply src/widgets/modals/StocktakeBatchModal.js]: Nice - definitely need to come up with a better idea as this clearly isn't working well

**Pull request merged**